Implement collateral contract

ColePBryan opened the pull request.

ColePBryan committed:
feat(collateral-contract): implements collateral contract

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
docs(collateral-contract): Adds documenting comments

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
style: Fix linting errors and warnings

fixes linting issues found with slither, only remaining linting issue is versions which couldn't be updated due to dependency version mismatch.

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
fix(collateral-contract): collateral account not being reduced on redeem

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
docs(collateral-contract): Adds revert causes and other info documentation

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
feat(collateral-contract): Adds external allocation id tracker

Allocation ID tracker is being added externally to allow for contract updates and changes without losing list of used allocation IDs.

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

aasseman commented on src/Collateral.sol:
```diff
@@ -0,0 +1,192 @@
+
+        emit Redeem(msg.sender, amount);
+        allocationIDTracker.useAllocationID(signedRAV.rav.allocationId);
+        require(collateralToken.transfer(msg.sender, amount));
```
I suppose that line (and perhaps more) will have to be changed to send to The Graph's rebate pool.

ColePBryan replied:
I added this as issue #6.

ColePBryan commented on src/Collateral.sol:
```diff
@@ -0,0 +1,192 @@
+    struct CollateralAccount {
+        uint256 balance; // Total collateral balance for a sender-receiver pair
+        uint256 amountThawing; // Amount of collateral currently being thawed
+        uint256 thawEnd; // Block number at which thawing period ends
```
The notion page says to use blocks: https://www.notion.so/edgeandnode/Collateralization-308e5652e4d644fd8319c80503d85400?pvs=4#356b57f2cacc44a891cd0bf50950417d

Could you discuss this in the notion page to hopefully get it updated there? That way the contract and the notion page stay in lockstep.

ColePBryan replied:
Rolling this into an issue #11 due to the scope creep on this PR.

ColePBryan commented on src/Collateral.sol:
```diff
@@ -0,0 +1,192 @@
+        address sender = authorizedSigners[signer];
+        address receiver = msg.sender;
+        uint256 amount = signedRAV.rav.valueAggregate;
+        require(collateralAccounts[sender][receiver].balance >= amount, "Insufficient collateral balance");
```
One issue with that is how do we handle collecting the remaining amount? Do we only allow one redeem? Do we allow gateways to have a debt that can be grabbed by the indexer after the collateral is refilled? I think we should create an issue for this and start the discussion, but I'm not sure this functionality will go in this PR.

aasseman replied:
Moved to #5

ColePBryan commented on src/Collateral.sol:
```diff
@@ -0,0 +1,192 @@
+        address signer = tapVerifier.recoverRAVSigner(signedRAV);
+        require(authorizedSigners[signer] != address(0), "Signer not authorized");
+
+        address sender = authorizedSigners[signer];
```
Would allowing them to get the collateral account using the signer work?

aasseman replied:
Yup that's what I had in mind

ColePBryan replied:
Check the latest commit. That should resolve this, but let me know if there is an issue.

aasseman replied:
I suppose getting the `sender` address is a YAGNI, so that should cover the most important piece of info the receiver needs.

ColePBryan committed:
feat(collateral-contract): get collateral account info using address of an authorized signer

adds a function that takes in a signer and receiver. Then matches the signer to the sender they are authorized to sign for. Finally, return the corresponding collateral account info for the  sender-receiver pair. This should allow a receiver to check available collateral based on a recovered address from a signature without specifically knowing the address of the sender.

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

tmigone commented on src/AllocationIDTracker.sol:
```diff
@@ -0,0 +1,51 @@
+     * @param allocationID The allocation ID to mark as used.
+     * @notice REVERT: This function may revert if the allocation ID has already been used.
+     */
+    function useAllocationID(address allocationID) public {
```
This being permissionless allows anyone to mark allocations as already used, potentially denying fees to everyone by front running txs. 


One way of protecting these methods is to require the caller to post a proof that they own the key pair behind an allocation id (see [this](https://github.com/613f824f0c3a43be9b4d07fa5e297ff0?pvs=25) snippet).

tmigone commented on src/AllocationIDTracker.sol:
```diff
@@ -0,0 +1,51 @@
+     * @param allocationIDs The allocation IDs to mark as used.
+     * @notice REVERT: This function may revert if any of the allocation IDs have already been used.
+     */
+    function useAllocationIDs(address[] memory allocationIDs) public {
```
Same concern with permissionless calls as with `useAllocationIDs`.

tmigone replied:
Solidity allows function overloading, so you could do this if you preferred: 

```
    function useAllocationID(address allocationIDs) public 

    function useAllocationID(address[] memory allocationIDs) public 
```

tmigone replied:
Could be `external`

tmigone replied:
`memory` could be `calldata`

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,202 @@
+    /**
+     * @dev Emitted when collateral is redeemed by a receiver.
+     */
+    event Redeem(address indexed receiver, uint256 amount);
```
We should add `allocationId` and `sender` (indexed) to this event

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,202 @@
+    /**
+     * @dev Emitted when a thaw request is made for collateral.
+     */
+    event ThawRequest(address indexed sender, address indexed receiver, uint256 amount, uint256 thawEnd);
```
Maybe add the total amount that ends up thawed (`account.amountThawing`)

tmigone replied:
Also I would rename the event to just `Thaw`

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,202 @@
+     */
+    event AuthorizeSigner(address indexed signer, address indexed sender);
+
+    constructor(address collateralToken_, address tapVerifier_, address allocationIDTracker_, uint256 thawingPeriod_) {
```
Total nit but we usually use leading `_`s so `_collateralToken` instead of `collateralToken_`.

tmigone replied:
Ah that's excellent. Looks like it's us who are out of spec then 😵‍💫

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,202 @@
+    mapping(address => address) private authorizedSigners;
+
+    // The ERC20 token used for collateral
+    IERC20 public immutable collateralToken;
```
Not all ERC20 tokens return a boolean on their transfer functions. These tokens would not be compatible with this version of the contract (there are several instances of `require(collateralToken.transfer(msg.sender, amount));` for example.


We could use OpenZeppelin’s Safe ERC20 helpers OR document the decision with a comment saying that it’s only intended to be used with GRT for example.

ColePBryan replied:
This is using the openzeppelin ERC20 interface which defines returning a bool for the transfer function. I would assume that any contract implementing that interface would have to adhere to the interface, right?

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/7cc2cbfeb56a00bfe055ebe0d27da902d95ed3d3/contracts/token/ERC20/IERC20.sol#L41

tmigone replied:
What is using OZ's ERC20 interface is the storage variable on the collateral contract but that does not say anything about what the actual target ERC20 contract used for it's implementation. 

You could initialize the collateral contract using the address of an ERC20 contract that does not adhere to OZ's interface, or even to the ERC20 spec. Using OZ's IERC20 on the collateral contract only tells the contract to consider an address as a contract that adheres to the IERC20 spec.

tmigone replied:
Not sure I made a good job at explaining this... perhaps a better way of explaining it is to think the IERC20 here is "casting" an address as "a contract that adheres to the ERC20 interface". You can "cast" any address but that does not mean that the address is an ERC20 contract.

ColePBryan replied:
It sounds like overall the chosen interface is simply a suggestion to the contract deployer. So if someone wanted to disregard the chosen interface there might not even be a `transfer` function?

ColePBryan replied:
So in this case it essentially doesn't matter what interface we use or what functions/returns. The deployer could use any address and it will get cast as the interface declared in the contract. It seems like technically they could just send in a random address pointing to nothing and the deploy might still work? I'm not sure if I see why changing this to some other interface would resolve that issue. Are you just saying that avoiding the `require` would likely allow a larger amount of ERC implementations to work since it would work whether or not it returns bool (assuming it always reverts on failure)?

tmigone replied:
> So in this case it essentially doesn't matter what interface we use or what functions/returns. The deployer could use any address and it will get cast as the interface declared in the contract. It seems like technically they could just send in a random address pointing to nothing and the deploy might still work?

Exactly. The IERC20 interface is only describing to this contract how you'll be interacting with the address you pass to it. If it's a random address the deploy will work but it will fail at "runtime" when a function call tries to interact with it.

> I'm not sure if I see why changing this to some other interface would resolve that issue. Are you just saying that avoiding the require would likely allow a larger amount of ERC implementations to work since it would work whether or not it returns bool (assuming it always reverts on failure)?

Exactly! The problem is that not all ERC20's revert on transfer failure and not all return true on success. The SafeERC20 interface is used to normalize this and just forget about it. This article has a nice explanation: https://soliditydeveloper.com/safe-erc20. 


To be honest if you **know** which are the ERC20's you'll be interacting with and you know how they behave then you can acknowledge the risk and not use SafeERC20 at all. Just add a comment in the code to document the decision, for example the [billing contract](https://github.com/edgeandnode/billing-contracts/blob/dfe0f0d9d4a4f1f56e065e2730a38a45df11dbac/contracts/Billing.sol#L20) does not use SafeERC20. [Here](https://github.com/edgeandnode/subscription-payments/blob/main/contract/audits/The%20Graph%20A-1%20_%20Macro%20Audits.pdf) is however another recent audit we had where we didn't specify this and the auditors reported it as a low issue (see L2).

ColePBryan replied:
Makes sense, thanks for the explanation. I think I'll switch to safe-erc20. I added this as issue #12  for tracking.

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,202 @@
+     * @notice The collateral must be approved for transfer by the sender.
+     * @notice REVERT: this function will revert if the collateral transfer fails.
+     */
+    function depositCollateral(address receiver, uint256 amount) external {
```
Consider renaming `depositCollateral` , `thawCollateral` and `withdrawThawedCollateral` to `deposit`, `thaw` and `withdraw`. IMO the meaning is implicit.

ColePBryan replied:
I think withdraw will be left slightly ambiguous (could be mistaken for redeem), but seems like this change will fit with existing naming style well enough.

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,202 @@
+    mapping(address => mapping(address => CollateralAccount)) private collateralAccounts;
+    // Map of authorized signers to which sender they are authorized to sign for
+    // key: signer address, value: sender address
+    mapping(address => address) private authorizedSigners;
```
If we upgrade solidity version to 0.8.18 we could used named parameters in mapping types which I think are great to improve code readability.

Note that this applies to any mapping but specially this one since it required an extra comment to explain what key and value are.

https://github.com/ethereum/solidity/blob/develop/Changelog.md#0818-2023-02-01

ColePBryan replied:
I am moving this to an issue to be addressed on a separate PR (#7 )

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,202 @@
+     */
+    function depositCollateral(address receiver, uint256 amount) external {
+        collateralAccounts[msg.sender][receiver].balance += amount;
+        emit Deposit(msg.sender, receiver, amount);
```
In general events should be emitted after all processing is done. This ensures a correct order of events in case there are interactions in the method call that also emit events. 

For example in this case the `Deposit` would be logged before the `Transfer` on the next line, which is not quite true since the deposit is only completed after the transfer is successful.

There are several instances of events being emitted before contract interactions in this contract.

ColePBryan replied:
I added this change but realized after why it was done in that order. I am using slither for linting and it gives the following detector: https://github.com/crytic/slither/wiki/Detector-Documentation#reentrancy-vulnerabilities-3

tmigone replied:
That's interesting, first time I see that recommendation. I understand where it's coming from, though I'd argue that you shouldn't be designing for well ordered re-entrancy events but rather what makes sense. 

To me a `Deposit` event (to cite an example) signals that a deposit has been finalized and that only happens _after_ the token transfer has been completed. So the event order should be `Transfer` then `Deposit`. But then again, this is probably something developers could start a war over 😆

ColePBryan replied:
Yeah I agree with your thought process. If the intent of the event was to show a deposit is being started then it would make sense to have it before, but this event is to show when it is being completed. I think ordering of events is more important than handling a re-entrancy that we are not expecting in the first place.

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,202 @@
+     * @dev Authorizes a signer to sign RAVs for the sender.
+     * @param signer Address of the authorized signer.
+     */
+    function authorizeSigner(address signer) external {
```
`authorizeSigner` being permissionless allows for a griefing attack where an attacker could block signer addresses by front-running calls and setting their own address as sender. 


Two potential solutions to this:
 1. Require proof of signer ownership before setting `authorizedSigners[signer]`, with a message signed by the signer, similar to what we suggested for the allocation id’s.
 2. Change the mapping to include the sender `authorizedSigners[sender][signer]`, however this would also require some way to retrieve the `sender` on the `Collateral.redeem()` call which is probably not practical.

ColePBryan replied:
Good catch. I think option 1 is probably the best option here.

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,202 @@
+
+        emit Redeem(msg.sender, amount);
+        allocationIDTracker.useAllocationID(signedRAV.rav.allocationId);
+        require(collateralToken.transfer(msg.sender, amount));
```
This design bypasses query rebate system. `Collateral.redeem()` should call `Staking.collect()` instead of directly transferring the GRT tokens to the indexer.

aasseman replied:
We are tracking this one in #6

tmigone commented on src/AllocationIDTracker.sol:
```diff
@@ -0,0 +1,51 @@
+ * @notice This contract is intended to be used with the `Collateral` contract.
+ */
+contract AllocationIDTracker {
+    mapping (address => bool) private _usedAllocationIDs;
```
Current `AllocationIDTracker` assumes a single gateway. We can future proof it by modifying the `_usedAllocationIDs` mapping to make the pair (allocation id, gateway) unique instead of just the allocation id. So maybe we can rename that contract to `RAVTracker` and make the mapping a mapping of a mapping where the first key is the gateway (sender on the redeem function) and the second key is the allocation id.

ColePBryan replied:
This will be tracked in #8

tmigone reviewed (changes requested):
🚀

ColePBryan committed:
feat(collateral-contract): switches thawing period to timestamp from number of blocks

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,202 @@
+     *                    collateral (greater than the amount in the RAV).
+     *                  - the allocation ID has already been used.
+     */
+    function redeem(TAPVerifier.SignedRAV memory signedRAV) external {
```
This can be calldata which should be cheaper:

```
function redeem(TAPVerifier.SignedRAV calldata signedRAV)
```

tmigone commented on src/TAPVerifier.sol:
```diff
@@ -33,34 +33,33 @@ contract TAPVerifier is EIP712 {
      */
-    function verifyRAVSignature(SignedRAV memory _signedRAV, address _address) public view returns (bool) {
-        return recoverRAVSigner(_signedRAV) == _address;
+    function verifyRAVSignature(SignedRAV memory signedRAV, address expectedAddress) public view returns (bool) {
```
This method can be `external`

tmigone commented on src/TAPVerifier.sol:
```diff
@@ -33,34 +33,33 @@ contract TAPVerifier is EIP712 {
-    function recoverRAVSigner(SignedRAV memory _signedRAV) public view returns (address) {
-        bytes32 messageHash = hashRAV(_signedRAV.rav);
-        return ECDSA.recover(messageHash, _signedRAV.signature);
+    function recoverRAVSigner(SignedRAV memory signedRAV) public view returns (address) {
```
All `memory` instances in this contract could be `calldata` and save a bit of gas.

tmigone commented on src/AllocationIDTracker.sol:
```diff
@@ -0,0 +1,51 @@
+     * @param allocationID The allocation ID to check.
+     * @return True if the allocation ID has been used, false otherwise.
+     */
+    function isAllocationIDUsed(address allocationID) public view returns (bool) {
```
This could be `external`

tmigone commented:
One additional QoL improvement we could do is to use solidity custom errors: https://blog.soliditylang.org/2021/04/21/custom-errors/

They make the code slightly more verbose but they bring improvements on error handling and tracing when a tx actually fails.

tmigone commented:
One last comment... I promise, linting seems to be broken so there are several lines that need linting (that's another easy QA classic from audits).

ColePBryan committed:
feat(collateral-contract): adds provable allocation ID ownership

allows keeping allocationIDTracker contract non governed while ensuring only owners of allocationID private key.

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
feat(collateral-contract): adds missing parameters to redeem event

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
style(collateral-contract): move event emission to after event is complete to maintain event ordering

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
style(collateral-contract): adds total amount thawing to thaw request event

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
style(collateral-contract): updating thaw request event name

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
style(collateral-contract): updates function names to reduce redundancy

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan committed:
refactor: reduce gas costs by using calldata over memory and external over public where possible

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

ColePBryan commented:
> One additional QoL improvement we could do is to use solidity custom errors: https://blog.soliditylang.org/2021/04/21/custom-errors/
> 
> They make the code slightly more verbose but they bring improvements on error handling and tracing when a tx actually fails.

#9

ColePBryan commented:
> One last comment... I promise, linting seems to be broken so there are several lines that need linting (that's another easy QA classic from audits).

What linter are you using? I am using slither right now and I haven't configured it much.

Feel free to comment on #10 with any suggestions for what is recommended.

ColePBryan committed:
fix(collater-contract): adds a proof for authorizing a signer to avoid griefing attack

The signer being authorized now has to provide the sender a proof that they are allowing themselves to be authorized. This is accomplished by the signer signing a keccak hash of the senders address.

BREAKING CHANGE:

changes api of authorizeSigner

Signed-off-by: Bryan Cole <bryan.cole@semiotic.ai>

aasseman reviewed (approved):
:shipit:

ColePBryan commented:
@tmigone I believe all comments you made have either been addressed with a code update or moved to an issue to be fixed after this PR. Could you confirm and approve the PR if that is the case?

tmigone commented on src/Collateral.sol:
```diff
@@ -0,0 +1,224 @@
+     * @dev Authorizes a signer to sign RAVs for the sender.
+     * @param signer Address of the authorized signer.
+     */
+    function authorizeSigner(address signer, bytes calldata proof) external {
```
I think the idea is to add a method to revoke an authorized signer right? If so then the proofs here allow for a replay attack, since the proof is public once posted you could re-authorize a signer even after the sender has revoked. 

We can fix that by adding a nonce to the signed message. The GRT contract implements a similar thing for permits here: https://github.com/graphprotocol/contracts/blob/dev/contracts/token/GraphToken.sol#L110
Essentially we add a mapping storing nonces per sender. Then part of the signed message is the current nonce. After validating the signed message the nonce gets incremented so the same signed message can not be re-used

ColePBryan replied:
I'm not sure if this is needed. The only one who would be able to revoke a signer is the sender. The reason for revoking a signer is if the sender feels the signer has been compromised. If that is the case what would be the reason for re-adding the signer? They could have just kept the signer in the first place.

tmigone replied:
For the sender there is no reason, but if I'm an attacker that got hold of signer keys I can reuse the proof that was originally submitted to re-add the signer as a valid one. Then I could forge my own RAVs and 🎄

ColePBryan replied:
Would an attacker be able to spoof `msg.sender` to be set to the senders address? Otherwise, I don't see how the attacker getting ahold of the proof would help them. The proof check uses the `msg.sender` for the call to authorizeSigner. They would only be able to add the signer as authorized to their own address.

ColePBryan replied:
This conversation can be continued on #18

tmigone replied:
Ahh good point. Yeah agreed then, there isn't much harm that could be done.

Even then, I just realized that the `verifyAuthorizedSignerProof` uses `msg.sender` as the message, so even with a stolen proof you couldn't get past the proof verification.

tmigone reviewed (changes requested):
Some comments on both proof uses.

ColePBryan commented on src/AllocationIDTracker.sol:
```diff
@@ -0,0 +1,57 @@
+     * @param allocationID The allocation ID to mark as used.
+     * @notice REVERT: This function may revert if the allocation ID has already been used.
+     */
+    function useAllocationID(address allocationID, bytes calldata proof) external {
+        require(!_usedAllocationIDs[allocationID], "Allocation ID already used");
+        require(verifyProof(proof, allocationID) == true, "Proof is not valid");
+        _usedAllocationIDs[allocationID] = true;
+        emit AllocationIDUsed(allocationID);
+    }
```
#17 We can continue the conversation here.

tmigone reviewed (approved):
🚀
